Automattic/woocommerce-payments
# Create UPE payment gateway

Closes #1826 

#### Changes proposed in this Pull Request


*Stripe Universal Payment Element Payment Gatway*


The changes in this PR contain the code to enable the UPE payment gateway, including the client functionality to complete a checkout. The UPE gateway is locked away behind a feature flag (`_wcpay_feature_upe`), which must be added to your store's `WC_Options` table. Once this flag is added, the UPE gateway will replace the default card gateway: the settings pages should remain the same, but at checkout the UPE should replace the credit card gateway.

Included in this PR is also the first piece of saving (card) payment methods during a UPE transaction. However, that will need to be completed in #1830, since the key `setup_future_usage` must be updated to `off_session` on the payment intent (this will ensure that the payment method is saved to the relevant Stripe customer upon confirmation of the payment intent).





#### Testing instructions


1. Please pull the changes in this PR and ensure you are running the latest version of the Payments Server.
2. Enter your WC Options table and add the feature flag `_wcpay_feature_upe` with a value of `1`.
3. Stripe UPE payment gateway should now replace the Credit Card and be enabled at checkout. Please go to **Payments > Settings** and ensure that WooCommerce Payments is enabled. 
4. Add an item from your shop to your cart and proceed through a regular checkout. At checkout, select "WooCommerce Payments" and you should see the UPE appear below the payment gateway. 
5. Complete a checkout using Stripe test credentials (or real ones if you fancy flexing). After hitting "Place Order", you should eventually be brought to the Order Confirmation screen.
6. Confirm that the order has been created correctly in **WooCommerce > Orders**. Confirm that charge details appear correctly  in **Payments > Transactions**.
7. Pat yourself on the back.



-------------------
